fastestimator/trace/io/model_eval.py

```diff
@@ -0,0 +1,446 @@
+        data_id: Data instance ID key. If provided, then sample-case test will return failing sample ID.
+    """
+    def __init__(self,
+                 test_cases: Union[TestCase, List[TestCase]],
+                 save_path: str,
+                 test_title: str = "Test",
+                 data_id: str = None) -> None:
+
+        self.test_title = test_title
+        self.test_cases = to_list(test_cases)
+        self.instance_cases = []
+        self.aggregate_cases = []
+        self.sample_id = data_id
+
+        all_inputs = set()
+        for case in self.test_cases:
+            all_inputs.update(case.criteria_inputs)
+            case.clean_result()
+            if case.aggregate:
+                self.aggregate_cases.append(case)
+            else:
+                self.instance_cases.append(case)
+
+        if self.sample_id:
+            all_inputs.update([data_id])
+
+        path = os.path.normpath(save_path)
+        path = os.path.abspath(path)
+        root_dir = os.path.dirname(path)
+        report = os.path.basename(path) or 'report'
+        report = report.split('.')[0]
+        self.save_dir = os.path.join(root_dir, report)
+        self.src_dir = os.path.join(self.save_dir, "resources")
+        os.makedirs(self.save_dir, exist_ok=True)
+        os.makedirs(self.src_dir, exist_ok=True)
+
+        super().__init__(inputs=all_inputs, mode="test")
+
+    def on_begin(self, data: Data) -> None:
+        self._sanitize_report_name()
+        self._initialize_json_summary()
+
+    def on_batch_end(self, data: Data) -> None:
+        for case in self.instance_cases:
+            result = case.criteria(*[data[var_name] for var_name in case.criteria_inputs])
+            if not isinstance(result, np.ndarray):
+                raise TypeError("Criteria return of per-instance test need to be ndarray with dtype bool")
+            elif result.dtype != np.dtype("bool"):
+                raise TypeError("Criteria return of per-instance test need to be ndarray with dtype bool")
+
+            result = result.reshape(-1)
+            case.result.append(result)
+            if self.sample_id:
+                data_id = to_number(data[self.sample_id]).reshape((-1, ))
+                if data_id.size != result.size:
+                    raise ValueError("Array size of criteria return doesn't match ID array size."
+                                     "Criteria return size should be equal to the batch_size that each entry represents"
+                                     "test result of corresponding data instance")
+                case.fail_id.append(data_id[result == False])
+
+    def on_epoch_end(self, data: Data) -> None:
+        for case in self.aggregate_cases:
+            result = case.criteria(*[data[var_name] for var_name in case.criteria_inputs])
+            if not isinstance(result, (bool, np.bool_)):
+                raise TypeError("criteria return of epoch-case test need to be bool")
```

Comment:
should let the user know which tests are giving the invalid response. also 'needs' rather than 'need'

Reply:
fixed